lxd/storage/drivers/drivers_lvm_utils.go

```diff
@@ -0,0 +1,693 @@
+	if d.volumeFilesystem(vol) == "btrfs" {
+		return "user_subvol_rm_allowed,discard"
+	}
+
+	return "discard"
+}
+
+// openLoopFile opens a loopback file and disable auto detach.
+func (d *lvm) openLoopFile(source string) (*os.File, error) {
+	if source == "" {
+		return nil, fmt.Errorf("No \"source\" property found for the storage pool")
+	}
+
+	if filepath.IsAbs(source) && !shared.IsBlockdevPath(source) {
+		unlock := locking.Lock(d.name, "", "")
+		defer unlock()
+
+		// Try to prepare new loop device.
+		loopF, err := PrepareLoopDev(source, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		// Make sure that LO_FLAGS_AUTOCLEAR is unset, so that the loopback device will not
+		// autodestruct on last close.
+		err = UnsetAutoclearOnLoopDev(int(loopF.Fd()))
+		if err != nil {
+			return nil, err
+		}
+
+		return loopF, nil
+	}
+
+	return nil, fmt.Errorf("Source is not loop file")
+}
+
+// isLVMNotFoundExitError checks whether the supplied error is an exit error from an LVM command
+// meaning that the object was not found. Returns true if it is (exit status 5) false if not.
+func (d *lvm) isLVMNotFoundExitError(err error) bool {
+	runErr, ok := err.(shared.RunError)
+	if ok {
+		exitError, ok := runErr.Err.(*exec.ExitError)
+		if ok {
+			waitStatus := exitError.Sys().(syscall.WaitStatus)
+			if waitStatus.ExitStatus() == 5 {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// pysicalVolumeExists checks if an LVM Physical Volume exists.
+func (d *lvm) pysicalVolumeExists(pvName string) (bool, error) {
+	_, err := shared.RunCommand("pvs", "--noheadings", "-o", "pv_name", pvName)
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return false, nil
+		}
+
+		return false, fmt.Errorf("Error checking for LVM physical volume \"%s\"", pvName)
+	}
+
+	return true, nil
+}
+
+// volumeGroupExists checks if an LVM Volume Group exists and returns any tags on that volume group.
+func (d *lvm) volumeGroupExists(vgName string) (bool, []string, error) {
+	output, err := shared.RunCommand("vgs", "--noheadings", "-o", "vg_tags", vgName)
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return false, nil, nil
+		}
+
+		return false, nil, fmt.Errorf("Error checking for LVM volume group \"%s\"", vgName)
+	}
+
+	output = strings.TrimSpace(output)
+	tags := strings.SplitN(output, ",", -1)
+
+	return true, tags, nil
+}
+
+// volumeGroupExtentSize gets the volume group's physical extent size in bytes.
+func (d *lvm) volumeGroupExtentSize(vgName string) (int64, error) {
+	output, err := shared.RunCommand("vgs", "--noheadings", "--nosuffix", "--units", "b", "-o", "vg_extent_size", vgName)
+	if err != nil {
+		if d.isLVMNotFoundExitError(err) {
+			return -1, errLVMNotFound
+		}
+
+		return -1, err
+	}
+
+	output = strings.TrimSpace(output)
+	return strconv.ParseInt(output, 10, 64)
+}
+
+// renameVolumeGroup renames a volume group.
+func (d *lvm) renameVolumeGroup(vgName, newVgName string) error {
```

Comment:
Only has a single caller, any reason to keep it separate?

Reply:
No, I'll merge this into the calling function.